src/app/SubscriptionResumptionStorage.h

```diff
@@ -0,0 +1,156 @@
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Interaction Model Event Generatorr Delegate.
+ *
+ */
+
+#pragma once
+
+#include <app/ReadClient.h>
+#include <lib/core/CHIPCore.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * A SubscriptionPersistenceDelegate is used to persist subscriptions when they are established.
+ *
+ * Allows application to append any type of TLV data as part of an event log entry. Events
+ * have a standard header applicable to all events and this class provides the
+ * ability to add additional data past such standard header.
+ */
+class SubscriptionResumptionStorage
+{
+public:
+    // TODO: Create RAII ScopedMemoryBuffer replacement container that does not require is_trivial elements
+
+    // Structs to hold
+    struct AttributePathParamsValues
+    {
+        EndpointId mEndpointId;   // uint16
+        ClusterId mClusterId;     // uint32
+        AttributeId mAttributeId; // uint32
+        void SetValues(AttributePathParams & params)
+        {
+            mEndpointId  = params.mEndpointId;
+            mClusterId   = params.mClusterId;
+            mAttributeId = params.mAttributeId;
+        }
+        AttributePathParams GetParams() { return AttributePathParams(mEndpointId, mClusterId, mAttributeId); }
+    };
+    struct EventPathParamsValues
+    {
+        EndpointId mEndpointId; // uint16
+        ClusterId mClusterId;   // uint32
+        EventId mEventId;       // uint32
+        bool mIsUrgentEvent;    // uint8
+        void SetValues(EventPathParams & params)
+        {
+            mEndpointId    = params.mEndpointId;
+            mClusterId     = params.mClusterId;
+            mEventId       = params.mEventId;
+            mIsUrgentEvent = params.mIsUrgentEvent;
+        }
+        EventPathParams GetParams() { return EventPathParams(mEndpointId, mClusterId, mEventId, mIsUrgentEvent); }
+    };
+
+    /**
+     * Struct to hold information about subscriptions
+     */
+    struct SubscriptionInfo
+    {
+        NodeId mNodeId;
+        FabricIndex mFabricIndex;
+        SubscriptionId mSubscriptionId;
+        uint16_t mMinInterval;
+        uint16_t mMaxInterval;
+        bool mFabricFiltered;
+        Platform::ScopedMemoryBufferWithSize<AttributePathParamsValues> mAttributePaths;
+        Platform::ScopedMemoryBufferWithSize<EventPathParamsValues> mEventPaths;
+    };
+
+    /**
+     * Struct to hold list of subscriptions
+     */
+    struct SubscriptionList
+    {
+        size_t mSize;
+        std::unique_ptr<SubscriptionInfo[]> mSubscriptions = std::make_unique<SubscriptionInfo[]>(CHIP_IM_MAX_NUM_SUBSCRIPTIONS);
```

Comment:
You cannot call side-effecting allocation in static initialization. It may cause crashes if any such objects are created before allocator init. It better to manage them at runtime

Reply:
We also often use platform alloc instead of direct heap. 
Is there precedent for using heap directly in client devices? I believe there are devices which do NOT have heap at all (platform memory is a static buffer and is used heap-like, but std heap does not exist)

Reply:
The API has changed to use iterator pattern and this is no longer there.